File: Class 45/cses2195.java
import java.io.*;
import java.util.*;

public class cses2195{
  public static void main(String[] args) throws IOException{
    BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
  
    int N = Integer.parseInt(in.readLine());
    Point[] pt = new Point[N], upper = new Point[N], lower = new Point[N];
    String[] tokens;
    
    for(int i = 0; i < N; i++){
      tokens = in.readLine().split(" ");
      int x = Integer.parseInt(tokens[0]), y = Integer.parseInt(tokens[1]);
      pt[i] = new Point(x, y);
    }
    
    Arrays.sort(pt, new Comparator<Point>(){
      public int compare(Point a, Point b){
        if(a.x == b.x) return a.y - b.y;
        return a.x - b.x;
      }
    });
    
    Point left = pt[0], right = pt[N - 1];
    int Usize = 0, Lsize = 0;
    
    for(int i = 0; i < N; i++){
      long cp = cross(left, right, pt[i]);
      
      if(cp >= 0){
        while(Usize >= 2 && cross(upper[Usize - 2], upper[Usize - 1], pt[i]) > 0)
          --Usize;
        upper[Usize] = pt[i];
        Usize += 1;
      }
      
      if(cp <= 0){
        while(Lsize >= 2 && cross(lower[Lsize - 2], lower[Lsize - 1], pt[i]) < 0)
          --Lsize;
        lower[Lsize] = pt[i];
        Lsize += 1;
      }
    }
    
    int total = Lsize + Usize - 2;
    System.out.println(total);
    
    for(int i = 0; i < Lsize; i++)
      System.out.println(lower[i].x + " " + lower[i].y);
    
    for(int i = Usize - 2; i > 0; i--)
      System.out.println(upper[i].x + " " + upper[i].y);
  }
  
  static Point minus(Point a, Point b){
    return new Point(a.x - b.x, a.y - b.y);
  }
  
  static long cross(Point a, Point b){
    return 1L * a.x * b.y - 1L * b.x * a.y;
  }
  
  static long cross(Point a, Point b, Point c){
    return cross(minus(b, a), minus(c, a));
  }
}

class Point{
  int x, y;
  public Point(int _x, int _y){ x = _x; y = _y; }
}

File: Class 45/cses2195.cpp
#ifdef LOCAL
	#define d(...) fprintf(stderr, __VA_ARGS__), fflush(stderr)
#else
	#define d(...)
#endif
#include <stdio.h>
#include <algorithm>
using namespace std;

typedef long long int lli;
struct point{
	int x, y;
};
const int MN = 2e5+5;
point pt[MN], upper[MN], lower[MN], hull[MN];
int N, LS, US, HS;

point operator - (point a, point b){
	return { a.x - b.x, a.y - b.y };
}

bool operator < (point a, point b){
	if(a.x == b.x) return a.y < b.y;
	return a.x < b.x;
}

lli cross(point a, point b){
	return 1LL * a.x * b.y - 1LL * b.x * a.y;
}

lli cross(point a, point b, point c){
	return cross(b - a, c - a);
}

int main(){
	scanf("%d", &N);
	for(int i = 0; i < N; i++){
		int x, y;
		scanf("%d %d", &x, &y);
		pt[i] = {x, y};
	}

	sort(pt, pt + N);
	point left = pt[0], right = pt[N - 1];

	upper[US++] = left;
	lower[LS++] = left;

	for(int i = 1; i < N; i++){
		lli cp = cross(left, right, pt[i]);
		if(cp >= 0){
			while(US >= 2 && cross(upper[US - 2], upper[US - 1], pt[i]) > 0)
			   --US;
			upper[US++] = pt[i];
		}
		if(cp <= 0){
			while(LS >= 2 && cross(lower[LS - 2], lower[LS - 1], pt[i]) < 0)
				--LS;
			lower[LS++] = pt[i];
		}
	}

	reverse(upper + 1, upper + US - 1);

	for(int i = 0; i < LS; i++)
		hull[HS++] = lower[i];
	for(int i = 1; i < US - 1; i++)
		hull[HS++] = upper[i];

	printf("%d\n", HS);
	for(int i = 0; i < HS; i++)
		printf("%d %d\n", hull[i].x, hull[i].y);
}
